tiledb/sm/filesystem/s3_thread_pool_executor.cc

```diff
@@ -41,12 +41,20 @@ namespace tiledb {
 namespace sm {
 
 S3ThreadPoolExecutor::S3ThreadPoolExecutor(ThreadPool* const thread_pool)
-    : thread_pool_(thread_pool) {
+    : thread_pool_(thread_pool)
+    , state_(State::RUNNING) {
   assert(thread_pool_);
 }
 
 S3ThreadPoolExecutor::~S3ThreadPoolExecutor() {
-  std::unique_lock<std::mutex> lock_guard(tasks_lock_);
+  assert(state_ = State::STOPPED);
+  assert(tasks_.empty());
+}
+
+void S3ThreadPoolExecutor::Stop() {
```

Comment:
Do you want to make this return a `Status`? Then you can return an error status from `task->get()` if there was an error.